sdk/lib/conutils/pager.c

```diff
@@ -31,7 +32,188 @@
+}
+
+static BOOL
+ConPagerDefaultAction(IN PCON_PAGER Pager)
+{
+    PCTCH TextBuff = Pager->TextBuff;
+    DWORD ich = Pager->ich;
+    const DWORD cch = Pager->cch;
+    DWORD iColumn = Pager->iColumn;
+    DWORD iLine = Pager->iLine;
+    DWORD lineno = Pager->lineno;
+    const DWORD ScreenColumns = Pager->ScreenColumns;
+    const DWORD ScrollRows = Pager->ScrollRows;
+    LONG nTabWidth = Pager->nTabWidth;
+    DWORD ichLast = ich;
+    UINT nCodePage;
+    BOOL IsCJK;
+    UINT nWidthOfChar = 1;
+    BOOL IsDoubleWidthCharTrailing = FALSE;
+
+    if (ich >= cch)
+        return TRUE;
+
+    nCodePage = GetConsoleOutputCP();
+    IsCJK = IsCJKCodePage(nCodePage);
+
+    /* Normalize the tab width: if negative or too large,
+     * cap it to the number of columns. */
+    if (nTabWidth < 0)
+        nTabWidth = Pager->ScreenColumns - 1;
+    else
+        nTabWidth = min(nTabWidth, Pager->ScreenColumns - 1);
+
+    if (Pager->dwFlags & CON_PAGER_FLAG_EXPAND_TABS)
+    {
+ExpandTab:
+        while ((Pager->nSpacePending > 0) && (iLine < ScrollRows))
+        {
+            ConCallPagerLine(Pager, L" ", 1);
+            --(Pager->nSpacePending);
+            ++iColumn;
+            if (iColumn % ScreenColumns == 0)
+            {
+                if (!(Pager->dwFlags & CON_PAGER_FLAG_DONT_OUTPUT))
+                    ++iLine;
+            }
+        }
+    }
+
+    for (; ich < cch && iLine < ScrollRows; ++ich)
+    {
+        Pager->lineno = lineno;
+
+        /* TAB character */
+        if (TextBuff[ich] == TEXT('\t') &&
+            (Pager->dwFlags & CON_PAGER_FLAG_EXPAND_TABS))
+        {
+            ConCallPagerLine(Pager, &TextBuff[ichLast], ich - ichLast);
+            if (nTabWidth == 0)
+            {
+                ichLast = ich + 1;
+                continue;
+            }
+            ichLast = ++ich;
+            Pager->nSpacePending += nTabWidth - (iColumn % nTabWidth);
+            goto ExpandTab;
+        }
+
+        /* NEWLINE character */
+        if (TextBuff[ich] == TEXT('\n'))
+        {
+            ConCallPagerLine(Pager, &TextBuff[ichLast], ich - ichLast + 1);
+            ichLast = ich + 1;
+            if (!(Pager->dwFlags & CON_PAGER_FLAG_DONT_OUTPUT))
+                ++iLine;
+            ++lineno;
+            iColumn = 0;
+            continue;
+        }
+
+        /* FORM-FEED character */
+        if (TextBuff[ich] == TEXT('\f') &&
+            (Pager->dwFlags & CON_PAGER_FLAG_EXPAND_FF))
+        {
+            /* Skip the form-feed and clear until the end of the screen */
+            ConCallPagerLine(Pager, &TextBuff[ichLast], ich - ichLast);
+            ichLast = ich + 1;
+            // FIXME: Should we handle CON_PAGER_FLAG_DONT_OUTPUT ?
+            while (iLine < ScrollRows)
+            {
+                ConCallPagerLine(Pager, L"\n", 1);
+                // CON_STREAM_WRITE(Pager->Screen->Stream, TEXT("\n"), 1);
+                // if (!(Pager->dwFlags & CON_PAGER_FLAG_DONT_OUTPUT))
```

Comment:
Same: Should we use CON_STREAM_WRITE or ConCallPagerLine ? (Here the aim when expanding the form-feed character is to print out newlines for "filling" the screen up to the bottom)

Reply:
For the future that extends the pager flags, we'd better not to use `CON_STREAM_WRITE` here directly.

Reply:
`// if (!(Pager->dwFlags & CON_PAGER_FLAG_DONT_OUTPUT))`
Uncomment this.

Reply:
If this gets uncommented there's the risk of not increasing iLine and therefore entering an infinite while loop.

Reply:
```c
for (n = 0; n < ScrollRows - iLine; ++n)
{
    ConCallPagerLine(Pager, L"\n", 1);
    if (!(Pager->dwFlags & CON_PAGER_FLAG_DONT_OUTPUT))
        ++iLine;
}
iColumn = 0;
continue;
```

Reply:
I'll check this; nevertheless now this looks slightly "overengineered" (just to output consecutive newlines).